// graphql-lambda/dist/SubscriptionManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionManager = void 0;
const iterall_1 = require("iterall");
// polyfill Symbol.asyncIterator
if (Symbol.asyncIterator === undefined) {
    Symbol.asyncIterator = Symbol.for('asyncIterator');
}
class SubscriptionManager {
    constructor({ getSubscriptionNameFromEvent = (event) => event.event, getSubscriptionNameFromConnection = (name) => name, subscriptionManagerStorage } = {}) {
        this.subscribersByEvent = (event) => {
            console.log("🗣 subscriptionManager.subscribersByEvent()", { event });
            return {
                [Symbol.asyncIterator]: () => {
                    const name = this.getSubscriptionNameFromEvent(event);
                    const subscriptions = this.subscriptions.get(name) || [];
                    const subscribers = subscriptions.filter((subscriber) => subscriber.event === name);
                    console.log({ subscribers });
                    return iterall_1.createAsyncIterator([subscribers]);
                },
            };
        };
        this.subscribe = async (names, connection, operation) => {
            console.log("🗣 subscriptionManager.subscribe()", { names, connection, operation });
            names.forEach((n) => {
                const name = this.getSubscriptionNameFromConnection(n, connection);
                const subscriptions = this.subscriptions.get(name);
                const subscription = {
                    connection,
                    operation,
                    event: name,
                    operationId: operation.operationId,
                };
                if (subscriptions == null) {
                    this.subscriptions.set(name, [subscription]);
                }
                else if (!subscriptions.find((s) => s.connection.id === connection.id)) {
                    subscriptions.push({
                        connection,
                        operation,
                        event: name,
                        operationId: operation.operationId,
                    });
                }
            });
        };
        this.unsubscribe = async (subscriber) => {
            const subscriptions = this.subscriptions.get(subscriber.event);
            console.log("🗣 subscriptionManager.unsubscribe", { subscriber });
            if (subscriptions) {
                this.subscriptions.set(subscriber.event, subscriptions.filter((s) => s.connection.id !== subscriber.connection.id));
            }
        };
        this.unsubscribeOperation = async (connectionId, operationId) => {
            console.log("🗣 subscriptionManager.unsubscribeOperation", { connectionId, operationId });
            this.subscriptions.forEach((subscribers, event) => {
                this.subscriptions.set(event, subscribers.filter((subscriber) => subscriber.connection.id !== connectionId &&
                    subscriber.operationId !== operationId));
            });
        };
        this.unsubscribeAllByConnectionId = (connectionId) => {
            console.log("🗣 subscriptionManager.unsubscribeAllByConnectionId", { connectionId });
            for (const key of this.subscriptions.keys()) {
                this.subscriptions.set(key, this.subscriptions
                    .get(key)
                    .filter((s) => s.connection.id === connectionId));
            }
            return Promise.resolve();
        };
        this.subscriptions = subscriptionManagerStorage || new Map();
        this.getSubscriptionNameFromEvent = getSubscriptionNameFromEvent;
        this.getSubscriptionNameFromConnection = getSubscriptionNameFromConnection;
    }
}
exports.SubscriptionManager = SubscriptionManager;
